src/main/kotlin/org/jetbrains/research/testspark/tools/llm/generation/hf/HuggingFaceRequestManager.kt

```diff
@@ -0,0 +1,89 @@
+import java.net.HttpURLConnection
+
+/**
+ * A class to manage requests sent to large language models hosted on HuggingFace
+ */
+class HuggingFaceRequestManager(project: Project) : IJRequestManager(project) {
+    private val url = "https://api-inference.huggingface.co/models/meta-llama/"
+    private val systemPrompt = "You are a helpful and honest code and programming assistant." +
+        " Please, respond concisely and truthfully."
+
+    // TODO: The user should be able to change these numbers in the plugin's settings
+    private val topProbability = 0.9
+    private val temperature = 0.9
+
+    private val llmErrorManager = LLMErrorManager()
+
+    override fun send(
+        prompt: String,
+        indicator: CustomProgressIndicator,
+        testsAssembler: TestsAssembler,
+        errorMonitor: ErrorMonitor,
+    ): SendResult {
+        val httpRequest = HttpRequests.post(
+            url + SettingsArguments(project).getModel(),
+            "application/json",
+        ).tuner {
+            it.setRequestProperty("Authorization", "Bearer $token")
+        }
+
+        // Add system prompt
+        if (chatHistory.size == 1) {
+            chatHistory[0] = ChatMessage(chatHistory[0].role, formatPrompt(systemPrompt, chatHistory[0].content))
+        }
+
+        val llmRequestBody = HuggingFaceRequestBody(chatHistory, Parameters(topProbability, temperature)).toMap()
+        var sendResult = SendResult.OK
+        try {
+            httpRequest.connect {
+                it.write(GsonBuilder().disableHtmlEscaping().create().toJson(llmRequestBody))
+                when (val responseCode = (it.connection as HttpURLConnection).responseCode) {
+                    HttpURLConnection.HTTP_OK -> (testsAssembler as JUnitTestsAssembler).consumeHFRequest(it)
+                    HttpURLConnection.HTTP_INTERNAL_ERROR -> {
+                        llmErrorManager.errorProcess(
+                            LLMMessagesBundle.get("serverProblems"),
+                            project,
+                            errorMonitor,
+                        )
+                        sendResult = SendResult.OTHER
+                    }
+
+                    HttpURLConnection.HTTP_BAD_REQUEST -> {
+                        llmErrorManager.errorProcess(
+                            LLMMessagesBundle.get("hfServerError"),
+                            project,
+                            errorMonitor,
+                        )
+                        sendResult = SendResult.OTHER
+                    }
+                }
+            }
+        } catch (e: HttpStatusException) {
+            log.error { "Error in sending request: ${e.message}" }
+        }
+        return sendResult
+    }
+
+    /**
+     * Creates the required prompt for Llama models. For more details see:
+     * https://huggingface.co/blog/llama2#how-to-prompt-llama-2
+     */
+    private fun formatPrompt(systemPrompt: String, userMessage: String): String {
```

Comment:
`systemPrompt` will never change in a runtime; once you move it into the `LLMDefaults.properties` you may access it directly here. 

I suggest naming the method `craftInstructionPrompt` or `createInstructionPrompt`; `formatPrompt` implies a kind of generic text formatting detached from the LLM context.

Reply:
I understood that this prompt modification is only required for the `Llama` family. Since we may integrate other non-Llama models in the future, the implementation will have to be modified. Could you leave a `TODO` comment concerning this aspect?

Reply:
Added a TODO in https://github.com/JetBrains-Research/TestSpark/pull/212/commits/0b00e8db3c3c474e63c765b3864c2c7991b5fa4b.